### PR TITLE
67: Parallelize image uploads

### PR DESCRIPTION
**Pull request by NedReid**
## Description:

- Changes the image upload feature to always upload 5 at a time

## Screenshots:

<details>
<summary>Open/Close screenshots</summary>
N/A
</details>

## Merge checklist:

-   [x] I have reviewed my own code
-   [x] I have added screenshots of any UI changes
-   [x] I haven't added any packages that significantly increased the
        bundle size (or justified them if so)
-   [x] I have made sure any UI changes work on mobile
-   [x] I have ran `npm run lint` and there were no errors
-   [x] I have rebased on the latest version of main

### CODE REVIEW

File: frontend/components/admin/Photos.jsx [hunk 1]
```diff
@@ -48,50 +50,62 @@ class Photos extends React.Component {
...
-                image.type !== "image/tiff" &&
-                image.type !== "image/webp"
-            ) {
-                await this.setState({
-                    skipped: this.state.skipped + "Skipped " + image.name + "\n",
-                });
+            if (currentPromises.length < NO_PARALLEL_UPLOADS) {
+                currentPromises.push(this.uploadSingleImage(i));
             } else {
-                const filePromise = new Promise((resolve) => {
```
Comment (Pulindu1) [inline frontend/components/admin/Photos.jsx:71]: Can you explain what lines 71-91 did, and why it has been removed.

File: frontend/components/admin/Photos.jsx [hunk 2]
```diff
@@ -48,50 +50,62 @@ class Photos extends React.Component {
...
-                                skipped: this.state.skipped + "Failed " + image.name + " upload\n",
-                            });
-                        }
-                        resolve();
-                    };
-                    reader.readAsDataURL(image);
-                });
-
-                await filePromise;
+                const [promiseToRemove] = await Promise.race(
```
Comment (Pulindu1) [inline frontend/components/admin/Photos.jsx:100]: What does lines 100-104 do?

**Review by Pulindu1 — COMMENTED** [review]
The code is easy to understand, except for the sections highlighted. Could you please explain these to me, I'm happy to approve after that.

File: frontend/components/admin/Photos.jsx (hunk 2 above)
Comment (NedReid) [reply frontend/components/admin/Photos.jsx:102]: So basically:
- `Promise.race` returns the result of the first promise to return (in this case the promises are waiting for an image to upload)
- the promises are stored in a list of promises. we map the promises to return the the promise once completed, so we have the promise we want to remove after the .race is complete.
- We want to remove the first promise to complete from the list, so we splice `currentPRomises` to remove `promiseToRemove`. 
- Now we have fewer than 5 promises in `currentPromises`, so we start uploading the next image and push that to the list of promises.

In a single sentence, these lines wait for one of the currently uploading images to finish uploading, then starts uploading another image. This means that there are constantly 5 images uploading at a single time, and never more.

I've added a comment above to make the what these lines are doing slightly clear 😊

**Review by Pulindu1 — APPROVED** [review]
After reading the replies, I understand the code better. I'm happy to approve this.

**Pull request merged**